new/docs/design/01-new-cloudware-plans.md

```diff
@@ -0,0 +1,358 @@
+[Underware](https://github.com/alces-software/underware) (following various
+incoming changes to this) will often be used to create these templates.
+
+As such, most complex aspects of creating Cloudware templates will be beyond
+the scope of Cloudware itself, however Cloudware will need some simple
+templating functionality in order to support using the `value`s of `Output`s
+from previous `Deployment`s in later `Deployment`s. Two initial considerations
+for choosing this templating system are as follows:
+
+1. The Cloudware templating system should be able to unobtrusively live
+   alongside the Underware templating system. For example, it should be
+   possible for a template to be rendered by Underware to automate complex
+   aspects of constructing the template like looping through various values,
+   having various conditional sections depending on the current Underware
+   configuration, rendering partial templates within the main template etc.,
+   and then for this Underware-rendered template to also be rendered by
+   Cloudware, without special consideration needing to be made in either of
+   these stages to support the other stage (like escaping ERB tags etc.).
+
+2. The Cloudware templating system should be intentionally simplistic and just
+   needs to support the insertion of `Output` values in otherwise static
+   templates  - anything more complex than this is beyond the scope of
+   Cloudware (and is the responsibility of Underware or however else templates
+   are being created).
+
+Given these requirements, [Mustache](https://github.com/mustache/mustache) may
+be a good choice for Cloudware templates, as this is intentionally simple and
+should be able to independently live alongside Underware's ERB templates if
+needed.
+
+Initially, I believe the following should be sufficient features within
+Cloudware's templating system (and attempting anything beyond this should be an
+error):
+
+- `{{ deployment.SOME_DEPLOYMENT.outputs.SOME_OUTPUT }}`: inserts the value of
+  the `Output` with `name` `SOME_OUTPUT`, for `Deployment` with `name`
+  `SOME_DEPLOYMENT`, in the template. If the `Deployment` or the `Output`
+  cannot be found then this should be an error.
+
+- `{{ node.SOME_NODE.outputs.SOME_OUTPUT }}`: similarly, this inserts the value
+  of the given `Output` for the `Node` with `name` `SOME_NODE` in the template,
+  and an error should occur if either of these cannot be found.
+
+  One distinction between this and the `Deployment` method of inserting an
+  `Output` is that the `Node` output should be specified without the
+  `${node_name}__` prefix, i.e.  `{{ node.gpunode01.outputs.ip }}` will insert
+  the `value` of the `Output` with `name` `gpunode01__ip`, obtained when the
+  `gpunode01` `Node` was deployed.
+
+## `cloud deploy PLATFORM NAME TEMPLATE` command
+
+This command will deploy template at (relative or absolute) path `TEMPLATE`, on
+given cloud `PLATFORM`, and with given `NAME` (aside: in practise `PLATFORM`
```

Comment:
Typically the user will be working within a single provider and thus the original option flag probably makes more sense. Alternatively the different provides might be split into separate tools.